signac/core/collection_api.py

```diff
@@ -0,0 +1,446 @@
+                self.sync()
+            except BaseException:  # rollback
+                self._data = backup
+                raise
+        else:
+            raise ValueError("The data must be a mapping or None.")
+
+    @classmethod
+    def _validate_key(cls, key):
+        "Emit a warning or raise an exception if key is invalid. Returns key."
+        if isinstance(key, str):
+            if '.' in key:
+                from ..errors import InvalidKeyError
+                raise InvalidKeyError(
+                    "keys may not contain dots ('.'): {}".format(key))
+            else:
+                return key
+        elif isinstance(key, cls.VALID_KEY_TYPES):
+            return cls._validate_key(str(key))
+        else:
+            from ..errors import KeyTypeError
+            raise KeyTypeError(
+                "keys must be str, int, bool or None, not {}".format(type(key).__name__))
+
+    def __delitem__(self, item):
+        self.load()
+        del self._data[item]
+        self.sync()
+
+    def __setitem__(self, key, value):
+        self.load()
+        with self._suspend_sync():
+            self._data[self._validate_key(key)] = self.from_base(value, parent=self)
+        self.sync()
+
+    def __getitem__(self, key):
+        self.load()
+        return self._data[key]
+
+    def __iter__(self):
+        self.load()
+        return iter(self._data)
+
+    def __call__(self):
+        self.load()
+        return self.to_base()
+
+    def __eq__(self, other):
+        if isinstance(other, type(self)):
+            return self() == other()
+        else:
+            return self() == other
+
+    def __len__(self):
+        self.load()
+        return len(self._data)
+
+    def keys(self):
+        self.load()
+        return self._data.keys()
+
+    def values(self):
+        self.load()
+        return self.to_base().values()
+
+    def items(self):
+        self.load()
+        return self.to_base().items()
+
+    def get(self, key, default=None):
+        self.load()
+        return self._data.get(key, default)
+
+    def pop(self, key, default=None):
+        self.load()
+        ret = self._data.pop(key, default)
+        self.sync()
+        return ret
+
+    def popitem(self, key, default=None):
+        self.load()
+        ret = self._data.pop(key, default)
+        self.sync()
+        return ret
+
+    def clear(self):
+        self._data = {}
+        self.sync()
+
+    def update(self, mapping):
+        self.load()
+        with self._suspend_sync():
+            for key, value in mapping.items():
+                self[key] = self.from_base(value, self)
+        self.sync()
+
+    def setdefault(self, key, default=None):
+        self.load()
+        with self._suspend_sync():
+            ret = self._data.setdefault(key, self.from_base(default, self))
+        self.sync()
+        return ret
+
+
+class SyncedDict(_SyncedDict):
+
+    _PROTECTED_KEYS = ('_data', '_suspend_sync_', '_load', '_sync', '_parent')
```

Comment:
Should this be a module-level constant? I wonder if this creates memory space for the tuple with every instantiation of the object or just for the class.

Reply:
Another possibility would be to use `__slots__` to have a static allocation of attributes. It might be worth a try, but don't worry about it if it is difficult or presents new problems. https://book.pythontips.com/en/latest/__slots__magic.html

Reply:
I might retract my suggestion to use `__slots__`. I saw a comment on this webpage and maybe it is not a helpful idea: https://www.python-course.eu/python3_slots.php

> We mentioned in the beginning that slots are preventing a waste of space with objects. Since Python 3.3 this advantage is not as impressive any more. With Python 3.3 Key-Sharing Dictionaries are used for the storage of objects. The attributes of the instances are capable of sharing part of their internal storage between each other, i.e. the part which stores the keys and their corresponding hashes. This helps to reduce the memory consumption of programs, which create many instances of non-builtin types.

Reply:
This isn't necessary, anything declared in the class is class-wide and will be shared among instances. To a first approximation, classes are also just objects in the hierarchy of the instance, and attribute lookups should just traverse that and find the class-level instance if the attribute is not found in the object. No copies should be created (even aside from the extra key-sharing magic referred to).